File: tests/TestConnection.java
/*
 * Logistics API Platform - Test kết nối Java → SQL Server
 * Compile: javac -cp mssql-jdbc-12.8.1.jre11.jar TestConnection.java
 * Run:     java -cp ".;mssql-jdbc-12.8.1.jre11.jar" TestConnection
 */

import java.sql.*;

public class TestConnection {
    
    // Cấu hình kết nối - điều chỉnh theo môi trường
    private static final String SERVER = "localhost";
    private static final String DATABASE = "LogisticsDB";
    private static final String USERNAME = "sa";
    private static final String PASSWORD = "YourStrongPassword123!";
    
    private static final String CONNECTION_URL = String.format(
        "jdbc:sqlserver://%s:1433;databaseName=%s;user=%s;password=%s;encrypt=true;trustServerCertificate=true;",
        SERVER, DATABASE, USERNAME, PASSWORD
    );

    public static void main(String[] args) {
        System.out.println("==================================================");
        System.out.println("🔗 TEST KẾT NỐI JAVA → SQL SERVER");
        System.out.println("==================================================");
        System.out.printf("  Server:   %s%n", SERVER);
        System.out.printf("  Database: %s%n", DATABASE);
        System.out.println();

        try {
            // 1. Load JDBC Driver
            Class.forName("com.microsoft.sqlserver.jdbc.SQLServerDriver");
            System.out.println("✅ JDBC Driver loaded thành công");

            // 2. Kết nối
            System.out.println("🔌 Đang kết nối...");
            Connection conn = DriverManager.getConnection(CONNECTION_URL);
            System.out.println("✅ Kết nối thành công!");
            System.out.println();

            // 3. Kiểm tra phiên bản
            DatabaseMetaData metaData = conn.getMetaData();
            System.out.printf("📊 SQL Server: %s %s%n", 
                metaData.getDatabaseProductName(),
                metaData.getDatabaseProductVersion().split("\n")[0]);
            System.out.println();

            // 4. Kiểm tra các bảng
            System.out.println("📋 Bảng trong database:");
            String[] tableNames = {"Users", "Partners", "Areas", "ShippingRates", "Orders", "OrderStatusHistory"};
            
            Statement stmt = conn.createStatement();
            for (String table : tableNames) {
                try {
                    ResultSet rs = stmt.executeQuery("SELECT COUNT(*) FROM [" + table + "]");
                    rs.next();
                    int count = rs.getInt(1);
                    System.out.printf("   ✅ %-25s → %d bản ghi%n", table, count);
                    rs.close();
                } catch (SQLException e) {
                    System.out.printf("   ❌ %-25s → KHÔNG TÌM THẤY%n", table);
                }
            }
            System.out.println();

            // 5. Test truy vấn
            System.out.println("🧪 Test truy vấn mẫu:");
            try {
                ResultSet rs = stmt.executeQuery(
                    "SELECT tracking_code, status, shipping_fee FROM Orders"
                );
                while (rs.next()) {
                    System.out.printf("   📦 %s | %-12s | %,.0f VND%n",
                        rs.getString("tracking_code"),
                        rs.getString("status"),
                        rs.getDouble("shipping_fee"));
                }
                rs.close();
            } catch (SQLException e) {
                System.out.println("   ⚠️ Chưa có dữ liệu (chưa chạy seed_data.sql)");
            }
            System.out.println();

            stmt.close();
            conn.close();

            System.out.println("==================================================");
            System.out.println("✅ TẤT CẢ KIỂM TRA ĐỀU THÀNH CÔNG!");
            System.out.println("==================================================");

        } catch (ClassNotFoundException e) {
            System.out.println("❌ JDBC Driver không tìm thấy!");
            System.out.println("💡 Tải driver: https://learn.microsoft.com/en-us/sql/connect/jdbc/download-microsoft-jdbc-driver-for-sql-server");
        } catch (SQLException e) {
            System.out.println("❌ Lỗi kết nối: " + e.getMessage());
            System.out.println();
            System.out.println("💡 Gợi ý khắc phục:");
            System.out.println("   1. Kiểm tra SQL Server đang chạy");
            System.out.println("   2. Kiểm tra thông tin kết nối trong code");
            System.out.println("   3. Đảm bảo đã chạy init_database.sql");
            System.out.println("   4. Kiểm tra TCP/IP protocol đã bật trong SQL Server Configuration Manager");
        }
    }
}

File: tests/test_connection.py
"""
Script test kết nối Python → SQL Server
Chạy: python test_connection.py

Kiểm tra:
1. Kết nối pyodbc trực tiếp
2. Kiểm tra database LogisticsDB tồn tại
3. Kiểm tra các bảng đã tạo
4. Đếm số bản ghi trong mỗi bảng
"""
import sys
import os

# Thêm path để import config
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "backend"))

def test_pyodbc_connection():
    """Test kết nối bằng pyodbc"""
    try:
        import pyodbc
    except ImportError:
        print("❌ pyodbc chưa được cài đặt. Chạy: pip install pyodbc")
        return False

    from app.config import DevelopmentConfig
    config = DevelopmentConfig()

    print("=" * 50)
    print("🔗 TEST KẾT NỐI PYTHON → SQL SERVER")
    print("=" * 50)
    print(f"  Server:   {config.DB_SERVER}")
    print(f"  Database: {config.DB_NAME}")
    print(f"  Driver:   {config.DB_DRIVER}")
    print()

    # 1. Kiểm tra drivers có sẵn
    print("📋 ODBC Drivers có sẵn:")
    drivers = [d for d in pyodbc.drivers() if 'SQL Server' in d]
    for d in drivers:
        print(f"   ✅ {d}")
    if not drivers:
        print("   ❌ Không tìm thấy ODBC Driver cho SQL Server!")
        print("   💡 Cài đặt: https://learn.microsoft.com/en-us/sql/connect/odbc/download-odbc-driver-for-sql-server")
        return False
    print()

    # 2. Test kết nối
    print("🔌 Đang kết nối...")
    try:
        conn = pyodbc.connect(config.PYODBC_CONNECTION_STRING, timeout=10)
        cursor = conn.cursor()
        print("   ✅ Kết nối thành công!")
        print()

        # 3. Kiểm tra database version
        cursor.execute("SELECT @@VERSION")
        version = cursor.fetchone()[0]
        print(f"📊 SQL Server Version:")
        print(f"   {version.split(chr(10))[0]}")
        print()

        # 4. Kiểm tra các bảng
        print("📋 Bảng trong database:")
        cursor.execute("""
            SELECT TABLE_NAME 
            FROM INFORMATION_SCHEMA.TABLES 
            WHERE TABLE_TYPE = 'BASE TABLE'
            ORDER BY TABLE_NAME
        """)
        tables = [row[0] for row in cursor.fetchall()]
        
        expected_tables = ['Users', 'Partners', 'Areas', 'ShippingRates', 'Orders', 'OrderStatusHistory']
        
        for table in expected_tables:
            if table in tables:
                cursor.execute(f"SELECT COUNT(*) FROM [{table}]")
                count = cursor.fetchone()[0]
                print(f"   ✅ {table:<25} → {count} bản ghi")
            else:
                print(f"   ❌ {table:<25} → KHÔNG TÌM THẤY")
        print()

        # 5. Test truy vấn mẫu
        print("🧪 Test truy vấn mẫu:")
        try:
            cursor.execute("SELECT tracking_code, status, shipping_fee FROM Orders")
            rows = cursor.fetchall()
            for row in rows:
                print(f"   📦 {row[0]} | {row[1]:<12} | {row[2]:>10,.0f} VND")
        except Exception as e:
            print(f"   ⚠️ Chưa có dữ liệu đơn hàng (chưa chạy seed_data.sql)")
        print()

        cursor.close()
        conn.close()

        print("=" * 50)
        print("✅ TẤT CẢ KIỂM TRA ĐỀU THÀNH CÔNG!")
        print("=" * 50)
        return True

    except pyodbc.Error as e:
        print(f"   ❌ Lỗi kết nối: {e}")
        print()
        print("💡 Gợi ý khắc phục:")
        print("   1. Kiểm tra SQL Server đang chạy")
        print("   2. Kiểm tra thông tin trong file .env")
        print("   3. Đảm bảo đã chạy init_database.sql")
        print("   4. Kiểm tra SQL Server cho phép SQL Authentication")
        return False


if __name__ == "__main__":
    success = test_pyodbc_connection()
    sys.exit(0 if success else 1)
